source/common/protobuf/utility.cc

```diff
@@ -505,6 +506,79 @@ std::string MessageUtil::CodeEnumToString(ProtobufUtil::error::Code code) {
   }
 }
 
+namespace {
+
+// Recursive helper method for MessageUtil::redact() below.
+void redactInPlace(Protobuf::Message* message, bool ancestor_is_redacted) {
+  // If the message is an `Any`, we can't use any type-based introspection. Instead, we have to
+  // unpack it to its original type to redact it...
+  auto* any = dynamic_cast<ProtobufWkt::Any*>(message);
+  if (any != nullptr) {
+    // 1. Extract the type name from the URL. This is guaranteed to be the portion of the URL
+    // following the last '/' character.
+    const std::string& type_url = any->type_url();
+    const std::size_t last_slash = type_url.find_last_of('/');
+    const std::string type_name = type_url.substr(last_slash + 1);
+    const auto* descriptor =
+        Protobuf::DescriptorPool::generated_pool()->FindMessageTypeByName(type_name);
+
+    // 2. Instantiate a mutable message of the correct dynamic type, and unpack the `Any` into it.
+    Protobuf::DynamicMessageFactory dmf;
+    std::unique_ptr<Protobuf::Message> typed_message(dmf.GetPrototype(descriptor)->New());
+    any->UnpackTo(typed_message.get());
+
+    // 3. Redact the typed message and repack it into the original `Any`.
+    redactInPlace(typed_message.get(), ancestor_is_redacted);
+    any->PackFrom(*typed_message);
+
+    return;
+  }
+
+  // Otherwise, use reflection to traverse all populated fields of this message...
+  const auto* reflection = message->GetReflection();
+  std::vector<const Protobuf::FieldDescriptor*> field_descriptors;
+  reflection->ListFields(*message, &field_descriptors);
+  for (const auto* field_descriptor : field_descriptors) {
+    // Redact if this field or any of its ancestors have the `redacted` option set.
```

Comment:
Why not just redact concrete fields?

Reply:
Not sure what you mean by "concrete" fields, but maybe this will answer your question (or at least anticipate somebody else's question):

If you look at the declaration of `TlsCertificate`, you'll see that the `private_key` and `password` fields are of type `core.DataSource` rather than raw `string`. I couldn't change `core.DataSource`'s `inline_string` field to be redacted, since it's a general-use type, I had to apply it one step up the tree.

Reply:
I see. Fair enough, that makes sense, thanks.

Reply:
Maybe add some comments to this effect for posterity.